### PR TITLE
Add new location to search for cuda libdevice files

### PR DESCRIPTION
**Pull request by 9prady9**

### CODE REVIEW

File: src/backend/cuda/CMakeLists.txt [hunk 1]
```diff
@@ -334,17 +334,23 @@ SET(libdevice_bc "")
...
+          IF (EXISTS ${_libdevice_bc_file})
+              ADD_CUSTOM_COMMAND(
+                  OUTPUT "${_libdevice_bc_copy}"
+                  DEPENDS "${_libdevice_bc_file}"
+                  COMMAND ${CMAKE_COMMAND} -E copy "${_libdevice_bc_file}" "${_libdevice_bc_copy}")
+              LIST(APPEND libdevice_bc ${_libdevice_bc_copy})
+              ADD_DEFINITIONS(-D"__LIBDEVICE_COMPUTE_${libdevice_compute}")
+              BREAK()
+          ENDIF()
+      ENDFOREACH()
```
Comment (pavanky) [inline src/backend/cuda/CMakeLists.txt]: This is not the right way to do this. Does this happen when you install the toolkit using ubuntu packages? 

What is the value of `CUDA_NVVM_HOME` in your case?

Comment (9prady9) [inline src/backend/cuda/CMakeLists.txt]: Yes.

 `CUDA_NVVM_HOME` is not set, in fact `CUDA_HOME`/`CUDA_PATH` is also not set.

**Comment by shehzan10** [comment]
Can you paste the output of `cmake -LA | grep CUDA` here.

File: src/backend/cuda/CMakeLists.txt (hunk 1 above)
Comment (shehzan10) [inline src/backend/cuda/CMakeLists.txt]: @pavanky `CUDA_NVVM_HOME` is an internal variable and will not show up in the cache.

File: src/backend/cuda/CMakeLists.txt [hunk 2]
```diff
@@ -334,17 +334,23 @@ SET(libdevice_bc "")
 IF (USE_LIBDEVICE)
   SET(libdevice_computes "")
   LIST(APPEND libdevice_computes "20" "30" "35" "50")
+  LIST(APPEND cuda_root_list "${CUDA_NVVM_HOME}")
+  # Below location of libdevice files when installed via package manager on Ubuntu 16.04
+  LIST(APPEND cuda_root_list "/usr/lib/nvidia-cuda-toolkit")
```
Comment (shehzan10) [inline src/backend/cuda/CMakeLists.txt]: Using paths like this won't work on Windows. It will error out.

**Comment by 9prady9** [comment]
```
CMake Warning at src/backend/cpu/CMakeLists.txt:75 (MESSAGE):
  LAPACK not found.  Functionality will be disabled


-- CUDA Compute Detection Output: 30
-- CUDA Compute Detection Return: 0
-- CUDA Compute Detection Worked
-- CUDA cusolver library available in CUDA Version 7.5
-- LIBDEVICE not found on system. CUDA JIT may be slower
CMake Warning at src/backend/opencl/CMakeLists.txt:29 (MESSAGE):
  LAPACK not found.  Functionality will be disabled


-- TESTS: CUDA backend is ON.
-- EXAMPLES: CUDA backend is ON.
BUILD_CUDA:BOOL=ON
CUDA_64_BIT_DEVICE_CODE:BOOL=ON
CUDA_ATTACH_VS_BUILD_RULE_TO_CUDA_FILE:BOOL=ON
CUDA_BUILD_CUBIN:BOOL=OFF
CUDA_BUILD_EMULATION:BOOL=OFF
CUDA_COMPUTE_20:BOOL=OFF
CUDA_COMPUTE_30:BOOL=ON
CUDA_COMPUTE_32:BOOL=OFF
CUDA_COMPUTE_35:BOOL=OFF
CUDA_COMPUTE_37:BOOL=OFF
CUDA_COMPUTE_50:BOOL=OFF
CUDA_COMPUTE_52:BOOL=OFF
CUDA_COMPUTE_53:BOOL=OFF
CUDA_COMPUTE_60:BOOL=OFF
CUDA_COMPUTE_61:BOOL=OFF
CUDA_COMPUTE_62:BOOL=OFF
CUDA_COMPUTE_DETECT:BOOL=ON
CUDA_CUDART_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libcudart.so
CUDA_CUDA_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libcuda.so
CUDA_GENERATED_OUTPUT_DIR:PATH=
CUDA_HAVE_GPU:BOOL=TRUE
CUDA_HOST_COMPILATION_CPP:BOOL=ON
CUDA_HOST_COMPILER:FILEPATH=/usr/bin/cc
CUDA_NVCC_EXECUTABLE:FILEPATH=/usr/bin/nvcc
CUDA_NVCC_FLAGS:STRING=
CUDA_NVCC_FLAGS_DEBUG:STRING=
CUDA_NVCC_FLAGS_MINSIZEREL:STRING=
CUDA_NVCC_FLAGS_RELEASE:STRING=
CUDA_NVCC_FLAGS_RELWITHDEBINFO:STRING=
CUDA_PROPAGATE_HOST_FLAGS:BOOL=ON
CUDA_SDK_ROOT_DIR:PATH=CUDA_SDK_ROOT_DIR-NOTFOUND
CUDA_SEPARABLE_COMPILATION:BOOL=OFF
CUDA_TARGET_CPU_ARCH:STRING=
CUDA_TOOLKIT_INCLUDE:PATH=/usr/include
CUDA_TOOLKIT_ROOT_DIR:PATH=/usr
CUDA_TOOLKIT_TARGET_DIR:PATH=/usr
CUDA_USE_STATIC_CUDA_RUNTIME:BOOL=ON
CUDA_VERBOSE_BUILD:BOOL=OFF
CUDA_VERSION:STRING=7.5
CUDA_cublas_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libcublas.so
CUDA_cudart_static_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libcudart_static.a
CUDA_cufft_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libcufft.so
CUDA_cupti_LIBRARY:FILEPATH=CUDA_cupti_LIBRARY-NOTFOUND
CUDA_curand_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libcurand.so
CUDA_cusolver_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libcusolver.so
CUDA_cusparse_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libcusparse.so
CUDA_nppc_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libnppc.so
CUDA_nppi_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libnppi.so
CUDA_npps_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libnpps.so
CUDA_nvvm_INCLUDE_DIR:PATH=/usr/include
CUDA_nvvm_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libnvvm.so
CUDA_rt_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/librt.so
```

File: src/backend/cuda/CMakeLists.txt (hunk 2 above)
Comment (9prady9) [inline src/backend/cuda/CMakeLists.txt]: Doesn't cmake take care of slashes like the way it does for path hints we give for commands ?

Comment (9prady9) [inline src/backend/cuda/CMakeLists.txt]: this path is used by exists command in IF block

Comment (shehzan10) [inline src/backend/cuda/CMakeLists.txt]: At best, put it in a `if(UNIX AND NOT APPLE)`. Don't leave it open.

**Comment by 9prady9** [comment]
@shehzan10 `${CUDA_TOOLKIT_ROOT_DIR}` is set to `/usr` on my laptop(ubuntu 16.04), so unless i add `/lib/nvidia-cuda-toolkit` prefix it is not working.

**Comment by pavanky** [comment]
@9prady9 use a new option for `CUDA_LIBDEVICE_DIR`. Do not do what you are doing in this commit.

**Comment by 9prady9** [comment]
build arrayfire linux ci

File: src/backend/cuda/CMakeLists.txt [hunk 3]
```diff
@@ -7,6 +7,7 @@ INCLUDE(CLKernelToH)
 INCLUDE(FindNVVM)
 
 OPTION(USE_LIBDEVICE "Use libdevice for CUDA JIT" ON)
+SET(CUDA_LIBDEVICE_DIR "${CUDA_NVVM_HOME}" CACHE PATH "Path where libdevice compute files are located")
```
Comment (pavanky) [inline src/backend/cuda/CMakeLists.txt]: This should be set to `${CUDA_NVVM_HOME}/libdevice`

File: src/backend/cuda/CMakeLists.txt [hunk 4]
```diff
@@ -332,20 +333,20 @@ CL_KERNEL_TO_H(
...
-        DEPENDS "${_libdevice_bc_file}"
-        COMMAND ${CMAKE_COMMAND} -E copy "${_libdevice_bc_file}" "${_libdevice_bc_copy}")
-      LIST(APPEND libdevice_bc ${_libdevice_bc_copy})
-      ADD_DEFINITIONS(-D"__LIBDEVICE_COMPUTE_${libdevice_compute}")
-    ENDIF()
-  ENDFOREACH()
+    SET(libdevice_computes "")
+    LIST(APPEND libdevice_computes "20" "30" "35" "50")
+    FOREACH(libdevice_compute ${libdevice_computes})
+        SET(_libdevice_bc_file "${CUDA_LIBDEVICE_DIR}/libdevice/libdevice.compute_${libdevice_compute}.10.bc")
```
Comment (pavanky) [inline src/backend/cuda/CMakeLists.txt]: This should just be `${CUDA_LIBDEVICE_DIR}`

**Comment by 9prady9** [comment]
build arrayfire linux ci

**Pull request merged**